File: kth-smallest-element-in-a-bst/Solution230.java
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class Solution230 {

    // T: O(n), S: O(h)
    public int kthSmallest(TreeNode root, int k) {
        var list = new ArrayList<Integer>();
        inorder(root, list);
        return list.get(k - 1);
    }

    // In-order traversal of a BST returns the elements in sorted ascending order.
    private void inorder(TreeNode node, List<Integer> list) {
        if (node == null) {
            return;
        }
        inorder(node.left, list);
        list.add(node.val);
        inorder(node.right, list);
    }

    // T: O(h + k), S: O(h)
    public int kthSmallestIterative(TreeNode root, int k) {
        var stack = new Stack<TreeNode>();
        var curr = root;

        while (curr != null || !stack.isEmpty()) {
            // Push all left nodes
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }

            // Pop and process
            curr = stack.pop();
            k--;

            if (k == 0) {
                return curr.val;
            }

            // Move to right subtree
            curr = curr.right;
        }

        return -1;
    }

    public static void main(String[] args) {
        var sol = new Solution230();

        var root1 = new TreeNode(3);
        root1.left = new TreeNode(1);
        root1.left.right = new TreeNode(2);
        root1.right = new TreeNode(4);
        System.out.println(sol.kthSmallest(root1, 1)); // 1
        System.out.println(sol.kthSmallestIterative(root1, 1)); // 1

        var root2 = new TreeNode(5);
        root2.left = new TreeNode(3);
        root2.left.left = new TreeNode(2);
        root2.left.left.left = new TreeNode(1);
        root2.left.right = new TreeNode(4);
        root2.right = new TreeNode(6);
        System.out.println(sol.kthSmallest(root2, 3)); // 3
        System.out.println(sol.kthSmallestIterative(root2, 3)); // 3
    }

}

File: kth-smallest-element-in-a-bst/solution230.go
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// T: O(n), S: O(h)
func kthSmallest(root *TreeNode, k int) int {
	var list []int
	inorder(root, &list)
	return list[k-1]
}

func inorder(node *TreeNode, list *[]int) {
	if node == nil {
		return
	}
	inorder(node.Left, list)
	*list = append(*list, node.Val)
	inorder(node.Right, list)
}

// T: T: O(h + k), S: O(h)
func kthSmallestIterative(root *TreeNode, k int) int {
	stack := []*TreeNode{}
	curr := root

	for curr != nil || len(stack) > 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		}

		curr = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		k--

		if k == 0 {
			return curr.Val
		}

		curr = curr.Right
	}

	return -1
}

func main() {
	root1 := &TreeNode{Val: 3}
	root1.Left = &TreeNode{Val: 1}
	root1.Left.Right = &TreeNode{Val: 2}
	root1.Right = &TreeNode{Val: 4}
	fmt.Println(kthSmallest(root1, 1))          // 1
	fmt.Println(kthSmallestIterative(root1, 1)) // 1

	root2 := &TreeNode{Val: 5}
	root2.Left = &TreeNode{Val: 3}
	root2.Left.Left = &TreeNode{Val: 2}
	root2.Left.Left.Left = &TreeNode{Val: 1}
	root2.Left.Right = &TreeNode{Val: 4}
	root2.Right = &TreeNode{Val: 6}
	fmt.Println(kthSmallest(root2, 3))          // 3
	fmt.Println(kthSmallestIterative(root2, 3)) // 3
}
